cognite_synthetic_tags/tag_resolver.py

```diff
@@ -0,0 +1,333 @@
+    ...     "value_1": Tag("A1") * 2,
+    ...     "value_2": Tag("A1") * 3 + Tag("B2") * 1000 * Tag("B3"),
+    ...     "value_3": Tag("B2"),
+    ... }
+    >>> TagResolver(dummy_value_store).resolve(specs)
+    {'value_1': 2, 'value_2': 6003, 'value_3': 2}
+    """
+
+    def __init__(
+        self,
+        value_store: TagValueStoreT,
+        additional_operations=None,
+        **additional_stores: TagValueStoreT,
+    ):
+        self.context: TagResolverContextT = {}
+        self._default_store_key = "value_store"
+        self.data_stores: Dict[str, TagValueStoreT] = {
+            self._default_store_key: value_store,
+            **additional_stores,
+        }
+        self.operations = DEFAULT_OPERATIONS.copy()
+        self.operations.update(additional_operations or {})
+        self._recursive_tags: List[Set[str]] = [set()]
+        self._specs: TagSpecsT = {}
+
+    def resolve(self, specs: TagSpecsT) -> Dict[str, TagValueT]:
+        """
+        This is the heavy-lift method that does all the necessary steps:
+         * Go through the `specs` dict and figure out all the tags that
+           need to be queries from CDF.
+         * Handle any literal values in the specs (values that are not
+           instances of `Tag`, e.g. `2` in `Tag("foo") * 2`).
+         * Perform calculations according to tag formulas (e.g. actually do
+           the multiplication by 2 in the example above).
+        """
+        # extract any literal values from the specs:
+        specs, literals = self._extract_literals_from_specs(specs)
+
+        # add literals to context so that they can be used when resolving:
+        self.context.update(literals)
+
+        # recursively find all the actual CDF tags from the specs:
+        self._specs = specs
+        real_tags: Dict[str, Set[str]] = {
+            store_key: set() for store_key in self.data_stores.keys()
+        }
+        for key, tag in specs.items():
+            if tag.formula:
+                # dig into the formula recursively
+                tags = self._collect_tags_from_formula(key, tag.formula)
+                for store_key, store_tags in tags.items():
+                    real_tags[store_key].update(store_tags)
+            else:
+                # no formula, it's just a tag name
+                store_key = self._get_item_value_store(tag)
+                real_tags[store_key].add(tag.name)
+                # TODO this branch is present in _collect_tags_from_formula too.
+                #  Should refactor it so that both branches are handled there.
+
+        # remove known values from query these can be:
+        #  - literals, they are already in context,
+        #  - previously fetch or calculated tags (basically caching, only
+        #    relevant when calling `resolve` more then once with a single
+        #    `TagResolver` instance).
+        for store_key, store_tags in real_tags.items():
+            known_tags = store_tags & set(self.context.keys())
```

Comment:
Might be more readable 🤷 
```suggestion
            known_tags = store_tags.intersection(self.context)
```